packages/lit-html/src/directives/ref.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * @license
+ * Copyright (c) 2020 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {Directive, directive, nothing, ElementPart} from '../lit-html.js';
+
+/**
+ * Creates a new Ref object, which is container for a reference to an element.
+ */
+export const createRef = <T = Element>() => new Ref<T>();
+
+class Ref<T = Element> {
+  value?: T;
+
+  set(v: T | undefined): void {
+    this.value = v;
+  }
+}
+
+type RefOrCallback = Ref | ((el: Element | undefined) => void);
+
+class RefDirective extends Directive {
+  private _element?: Element;
+
+  render(_ref: RefOrCallback) {
+    return nothing;
+  }
+
+  update(part: ElementPart, [ref]: Parameters<this['render']>) {
+    if (typeof ref === 'function') {
+      if (this._element !== part.element) {
+        ref((this._element = part.element));
+      }
+    } else if (ref.value !== part.element) {
```

Comment:
If you pass a different ref object, it'll always be filled, but if you pass a different ref callback it won't. While a corner case, that's unexpected. Seems like we should either always call the ref function or store and dirty check it.

Reply:
Suggest we decide whether or not to support changing what you pass to `ref` and handling the callback/object cases the same.